src/rust/receiver-mock/src/router/otlp.rs

```diff
@@ -97,34 +84,320 @@ fn get_otlp_lines_from_logs(resource_logs: &logsv1::ResourceLogs) -> Vec<String>
-    let value = match &anyvalue.value {
-        Some(v) => v,
-        None => return String::new(),
-    };
-    let s = match value {
-        commonv1::any_value::Value::StringValue(inner) => inner.clone(),
-        commonv1::any_value::Value::BoolValue(inner) => inner.to_string(),
-        commonv1::any_value::Value::IntValue(inner) => inner.to_string(),
-        commonv1::any_value::Value::DoubleValue(inner) => inner.to_string(),
-        _ => String::new(),
-    };
+pub async fn handler_receiver_otlp_metrics(
+    req: HttpRequest,
+    body: web::Bytes,
+    app_state: web::Data<AppState>,
+    opts: web::Data<options::Options>,
+) -> impl Responder {
+    let remote_address = get_address(&req);
+
+    let content_type = get_content_type(&req);
+
+    if let Some(response) = try_dropping_data(&opts, &content_type) {
+        return response;
+    }
+
+    match content_type.as_str() {
+        OTLP_PROTOBUF_FORMAT_CONTENT_TYPE => {
+            let metrics_data: metricsv1::MetricsData = match metricsv1::MetricsData::decode(&mut Cursor::new(body)) {
+                Ok(data) => data,
+                Err(_) => return HttpResponse::BadRequest().body("Unable to parse body"),
+            };
+            let mut result = MetricsHandleResult::new();
+
+            // TODO: Consider giving it some basic capacity to avoid too many allocations.
+            let mut lines = vec![];
+            for resource_metrics in metrics_data.resource_metrics {
+                let resource_attributes = &resource_metrics.resource.unwrap().attributes;
+                for instrumentation_lib_metrics in resource_metrics.instrumentation_library_metrics {
+                    for metric in instrumentation_lib_metrics.metrics {
+                        // Translating the metrics to prometheus format allows us to use the already existing API for parsing metrics into samples.
+                        let metric_str_vec = format::otlp_metric_to_prometheus_string(&metric, resource_attributes);
+
+                        for m in metric_str_vec {
+                            if opts.print.metrics {
+                                println!("metrics => {}", m);
+                            }
+                            if opts.store_metrics {
+                                lines.push(m);
+                            }
+                        }
+
+                        result.handle_metric(metric.name);
+                        result.handle_ip(remote_address);
+                    }
+                }
+            }
+
+            if opts.store_metrics {
+                result.metrics_samples = lines_to_samples(lines);
+            }
+
+            app_state.add_metrics_result(result, &opts);
+        }
+        &_ => {
+            return get_invalid_header_response(&content_type);
+        }
+    }
 
-    return s;
+    HttpResponse::Ok().body("")
 }
 
+mod format {
+    use std::collections::HashMap;
```

Comment:
I'm not against merging this, but it really looks like more work than just converting from OTLP to Sample directly.

Reply:
However, there's one thing to consider then: how to log received metrics? Rust by default supports debug formatting with `{:?}`, which simply prints the struct how it is, but we could also consider adding some pretty printing.

Reply:
Now that's probably a thing for another PR, eg. the one solving #293 - I was planning to do it anyway, so it's not a problem for me.
My suggestion is that I'll create a mapping from otlp structures to samples and use `println("metrics => {:?}", sample)` in this PR, and in another one I'll add all the other things, like a `sample_to_line` function.

Reply:
OK, basically this mapping is finished, but still needs to be properly tested and refactored. Nevertheless, it is ready to review (but basically it's the same as before, just with `Sample`s instead of `String`s)